### PR TITLE
Youtube Search

### PR DESCRIPTION
**Pull request by minkj1992**
## Change logs
- Add youtube client
	- #15 
	- #16 
	- #17 
	- #20 
- Added middlewares
	- `etag` middleware
	- `cors` middleware 
- Added swagger
- #45

## Details
- filter youtube playlist (set videoDuration only more than 20m)

## Tests
Screen Shot 2022-07-30 at 10 33 01 PM

> http://127.0.0.1:8080/api/v1/musics?order=count&q=상쾌,일본

```
{
  "code": 0,
  "data": [
    {
      "youtubeVideoID": "hz27Smhnv7I",
      "thumbnailURL": "https://i.ytimg.com/vi/hz27Smhnv7I/default.jpg",
      "title": "[Playlist] 안질리는 유명한 J-POP 모음│YOASOBI, ZUTOMAYO, 요루시카, 카미야마 요우, Eve",
      "youtubeTags": [
        "우타이테",
        "일본노래",
        "유명한일본노래"
      ],
      "videoDeepLink": "vnd.youtube://hz27Smhnv7I",
      "isMemoed": false
    },
    {
      "youtubeVideoID": "wGf1-8b8fZo",
      "thumbnailURL": "https://i.ytimg.com/vi/wGf1-8b8fZo/default.jpg",
      "title": "[ PLAYLIST ] #26 상쾌하고✨기분 좋아지는 일본 노래 (가사, 번역) | Cheerful Japanese Songs (lyrics) | 明るい歌 (歌詞)",
      "youtubeTags": [
        "아이묭",
        "슈퍼플라이",
        "수퍼플라이",
        "레이야스다",
        "야스다레이",
        "음악",
        "음악추천",
        "노래",
        "노래추천",
        "일본",
        "일본노래",
        "제이팝",
        "밤",
        "늦은밤",
        "밤늦게",
        "감성",
        "감성적인노래",
        "락음악",
        "락",
        "팝",
        "시티팝",
        "플레이리스트",
        "노래리스트",
        "LIST",
        "playlist",
        "emotional",
        "emotion",
        "music",
        "song",
        "lyrics",
        "가사",
        "번역",
        "통역",
        "통번역",
        "jpop",
        "japan",
        "japanese",
        "愛を伝えたいだとか",
        "Ai Wo Tsutaetaidatoka",
        "あいみょん",
        "Aimyon",
        "Rei Yasuda",
        "Ohara Sakurako",
        "Little Glee Monster",
        "Superfly",
        "Shishamo"
      ],
      "videoDeepLink": "vnd.youtube://wGf1-8b8fZo",
      "isMemoed": false
    },
    {
      "youtubeVideoID": "d5jMy9nF-3A",
      "thumbnailURL": "https://i.ytimg.com/vi/d5jMy9nF-3A/default.jpg",
      "title": "청량 치사량 🐬 일본노래 모음 [jpop playlist]",
      "youtubeTags": [
        "playlist",
        "플레이리스트",
        "jpop",
        "일본노래",
        "청량"
      ],
      "videoDeepLink": "vnd.youtube://d5jMy9nF-3A",
      "isMemoed": false
    },
	... 중략 ...
  ]
}
```

### CODE REVIEW

**Comment by minkj1992** [comment]
- [x] need swagger
- [x] filter more than 20min videos
- [x] order By
- [x] plural to singular
- [x] rename search to searchPlaylist() in youtibe client

**Comment by minkj1992** [comment]
- TODO
  - pagination
  - mongo cache (music list, music detail)

File: application/controller/index.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
+package controller
+
+import "github.com/google/wire"
+
+var Set = wire.NewSet(NewMusicController)
+
+type HTTPCode int32
+
+// https://github.com/googleapis/googleapis/blob/643311525df18aff3090d13449db4e9c2e9df419/google/rpc/code.proto#L32
+const (
```
Comment (minkj1992) [inline application/controller/index.go:10]: perm link

File: application/controller/index.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	// HTTP Mapping: 503 Service Unavailable
+	Unavailable
+
+	// Unrecoverable data loss or corruption.
+	//
+	// HTTP Mapping: 500 Internal Server Error
+	DataLoss
+)
+
+type BaseResponse struct {
```
Comment (minkj1992) [inline application/controller/index.go:166]: perm link

File: application/controller/index.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+var Set = wire.NewSet(NewMusicController)
+
+type HTTPCode int32
+
+// https://github.com/googleapis/googleapis/blob/643311525df18aff3090d13449db4e9c2e9df419/google/rpc/code.proto#L32
+const (
+	// Not an error; returned on success
+	//
+	// HTTP Mapping: 200 OK
+	Ok HTTPCode = iota
```
Comment (devjaewon) [inline application/controller/index.go:14]: Is the use of this const some kind of enum?

as i understand
```go
const Ok = 0
const Cancelled = 1
const Unknown = 2
...
```

but shouldn't the value be like below?
```go
const Ok = 200
const Cancelled = 499
const Unknown = 500
...
```

File: application/controller/index.go (hunk 2 above)
Comment (devjaewon) [reply application/controller/index.go:166]: good!! i need it. merge plz~~

**Review by devjaewon — APPROVED** [review]

File: application/controller/index.go (hunk 2 above)
Comment (mcauto) [reply application/controller/index.go:166]: I think the `base` prefix is an irresponsible expression. It does not describe any functionality.
Synthesis with words that do not add explanation causes cognitive load.

I think `JSONResponse` or `HTTPResponse` is better.

Why don't you use this? [net/http#response](https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/net/http/response.go;l=35)

Use that intention in the class name.

File: application/controller/index.go (hunk 3 above)
Comment (mcauto) [reply application/controller/index.go:14]: Why don't you use this? [net/http#status](https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/net/http/status.go;l=9)

**Pull request merged**